packages/pubsub/src/publisher.js

```diff
@@ -0,0 +1,146 @@
+/*!
+ * Copyright 2017 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*!
+ * @module pubsub/publisher
+ */
+
+'use strict';
+
+var arrify = require('arrify');
+var common = require('@google-cloud/common');
+var extend = require('extend');
+var is = require('is');
+var prop = require('propprop');
+
+/**
+ *
+ */
+function Publisher(topic, options) {
+  options = extend(true, {
+    batching: {
+      maxBytes: Math.pow(1024, 2) * 5,
+      maxMessages: 1000,
+      maxMilliseconds: 1000
+    }
+  }, options);
+
+  this.topic = topic;
+  this.api = topic.api;
+
+  this.inventory_ = {
+    queued: []
+  };
+
+  Object.defineProperty(this.inventory_, 'queueBytes', {
+    get: function() {
+      var size = 0;
+
+      this.queued.forEach(function(message) {
+        size += message.data.size;
+      });
+
+      return size;
+    }
+  });
+
+  this.settings = {
+    batching: {
+      maxBytes: Math.min(options.batching.maxBytes, Math.pow(1024, 2) * 9),
+      maxMessages: Math.min(options.batching.maxMessages, 1000),
+      maxMilliseconds: options.maxMilliseconds
+    }
+  };
+
+  this.timeoutHandle_ = null;
+}
+
+/**
+ *
+ */
+Publisher.prototype.publish = function(data, attrs, callback) {
+  if (!(data instanceof Buffer)) {
+    throw new TypeError('Data must be in the form of a Buffer.');
+  }
+
+  if (is.fn(attrs)) {
+    callback = attrs;
+    attrs = {};
+  }
+
+  var opts = this.settings.batching;
+
+  var hasMaxMessages = this.inventory_.queued.length >= opts.maxMessages;
+  var hasMaxBytes = (this.inventory_.queueBytes + data.size) >= opts.maxBytes;
+
+  if (hasMaxMessages || hasMaxBytes) {
+    clearTimeout(this.timeoutHandle_);
+    this.timeoutHandle_ = null;
+    this.publish_();
+  } else if (!this.timeoutHandle_) {
+    this.timeoutHandle_ = setTimeout(
+      this.publish_.bind(this), opts.maxMilliseconds);
+  }
+
+  this.inventory_.queued.push({
+    data: data,
+    attrs: attrs,
+    callback: callback
+  });
+};
+
+/**
+ * This should never be called directly.
+ */
+Publisher.prototype.publish_ = function() {
+  var self = this;
+
+  var messages = this.inventory_.queued;
+  var callbacks = messages.map(prop('callback'));
+
+  this.inventory_.queued = [];
+
+  var reqOpts = {
+    topic: this.topic.name,
+    messages: messages.map(function(message) {
```

Comment:
Actually this happens because I included the callback function supplied to `publish` with this object. I could instead manage 2 arrays - 1 for the payload and the other for all the callbacks - and we could do away with this.